Added fields to enhance mapping and filtering possibilities

boecklic opened the pull request:
Besides it's original intention (satellite data), STAC and STAC API proved to be useful for many other data having a georeference of some kind as well. For many of them the core specification of STAC and STAC API is sufficient. In case of weather forecast data, there are limitations however due to the vast amount of data, the frequent update cycles and the nature of the data. 

Weather forecast data is typically calculated or recalculated every X hours (e.g. 3h) and results conceptually in a large number of datapackages, each containing the result of a specific selection of parameters among the many degrees of freedom (spatial, temporal, others...). The fields offered by STAC core allow to distinguish spatial and temporal parameters, but not more. Mapping these > 100'000 datapackages to STAC would either mean few items with an extensive list of assets, or a large list of items with one or a few assets each, or combining several parameter combinations in one file and having large files (50GB +). Each of these solutions is not very user-friendly, since selecting a subset of the data is hard to impossible (E.g. the spatial extent is often the same for every datapackage (covering the whole weather model grid) so it's of limited use for e.g. filtering).

We propose to add additional fields to the forecast extension to enhance mapping and filtering possibilities for forecast data. The degrees of freedom the data contains are
- 2D extent of the model grid
- datetime when the data was calculated
- datetime in the future to which the forecast corresponds
- level (third spatial dimension, mostly just an index)
- physical parameter (temperature, pressure etc.)
- perturbation sample

Open questions:
- We have implemented the query extension (https://github.com/stac-api-extensions/query, https://data.geo.admin.ch/api/stac/static/spec/v0.9/apitransactional.html#tag/STAC/operation/postSearchSTAC) so basically we are able to filter on these attributes as well.
- `forecast:param` can be a `string` or a list or strings. Having just a string would mean each parameter has to be in a separate file, having a list of strings would allow to combine parameters in one file. Although a more practical scenario is probably rather to just have a few parameter groups, e.g. "surface" or "full atmosphere" parameter group which could be implemented with just a `string`. Having just `string`s and not lists is also easier for querying/filtering the data, so we propose to just have strings. The allowed values for `forecast:param` are intentionally unrestricted since these heavily depend on the forecast model configuration.
- As for perturbation: users are often either interested in either the control run or all of the perturbation runs. For this it would be sufficient to have a `forecast:mode` attribute with values of either `ctrl` or `perturb`. If desired/necessary it could be paired with a `forecast:sample_nr` with the number of the perturbation run. Since we'll not currently use this it can also be omitted for now and added later if there's a usecase for it.
- The `forecast:level` will also not be implemented/used in our usecase, but could be useful for others.

boecklic committed:
added fields to inhance mapping and filtering possibilities for forecast data

m-mohr commented:
Hey @zakjan, @cboettig, @aaronspring,
you were active when we defined the first version of this extension.
I'd love to hear your feedback as I don't really have an understanding of the details.

cboettig commented:
cc @rqthomas

aaronspring commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions).  **TODO: should this be either string or list of strings to allow for situations where several parameters are combined in a file?**|
+| forecast:mode                | string | Denotes whether the data corresponds to the control run or perturbed runs. Allowed values are `ctrl` or `perturb`. |
+| forecast:sample_nr           | integer | Specifies the sample number of perturbed runs, e.g. `1`. **TODO: this could be omitted for the moment and added when there's a concrete use case** |
```
What about "member"? Or realization as in the netcdf conventions https://cfconventions.org/cf-conventions/cf-conventions.html?utm_source=perplexity

m-mohr replied:
I think it would be nice to follow existing conventions such as CF, indeed.

boecklic replied:
@aaronspring @m-mohr thanks for the review and the suggestion. I'll rename the attribute to `forecast:member`.

aaronspring commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:reference_datetime  | string | **REQUIRED.** The *reference* datetime: i.e. predictions for times after this point occur in the future. Predictions prior to this time represent 'hindcasts', predicting states that have already occurred. This must be in UTC. It is formatted according to [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). |
 | forecast:horizon             | string | The time between the reference datetime and the forecast datetime. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT6H` for a 6-hour forecast. |
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions).  **TODO: should this be either string or list of strings to allow for situations where several parameters are combined in a file?**|
```
I wonder why we didn't have a variable name before. Probably it was mostly theoretical and no practical usecase yet

m-mohr replied:
I assume some implementations used the datacube extension to express variables.

m-mohr replied:
... and some probably just created https://github.com/stac-extensions/cf for the same usecase.

aaronspring commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions).  **TODO: should this be either string or list of strings to allow for situations where several parameters are combined in a file?**|
+| forecast:mode                | string | Denotes whether the data corresponds to the control run or perturbed runs. Allowed values are `ctrl` or `perturb`. |
+| forecast:sample_nr           | integer | Specifies the sample number of perturbed runs, e.g. `1`. **TODO: this could be omitted for the moment and added when there's a concrete use case** |
+| forecast:level               | integer | Index of the vertical level of the height coordinate system used in the forecast model, e.g. `4`. **TODO: this could be omitted for the moment**|
```
So we require level to be filled or would this be optional?

m-mohr replied:
According to the description this would be optional right now.

boecklic replied:
@m-mohr @aaronspring we propose to have it optional. We'll not need it at the moment (all vertical levels will be in one file), but think that it might be useful later to e.g. have the ground level in a separate file.

m-mohr commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:reference_datetime  | string | **REQUIRED.** The *reference* datetime: i.e. predictions for times after this point occur in the future. Predictions prior to this time represent 'hindcasts', predicting states that have already occurred. This must be in UTC. It is formatted according to [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). |
 | forecast:horizon             | string | The time between the reference datetime and the forecast datetime. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT6H` for a 6-hour forecast. |
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions). |
```
I'm wondering how this relates to https://github.com/stac-extensions/cf ?
Seems very similar. If it's the same or very similar, I'd hope we can align to have only one way of expressing this.

m-mohr commented:
> * `forecast:param` can be a `string` or a list or strings. Having just a string would mean each parameter has to be in a separate file, having a list of strings would allow to combine parameters in one file. Although a more practical scenario is probably rather to just have a few parameter groups, e.g. "surface" or "full atmosphere" parameter group which could be implemented with just a `string`. Having just `string`s and not lists is also easier for querying/filtering the data, so we propose to just have strings. The allowed values for `forecast:param` are intentionally unrestricted since these heavily depend on the forecast model configuration.

You can also use the parameter on the asset level if you have multiple assets with different parameters.

I'm wondering how this property relates to https://github.com/stac-extensions/cf ?
Seems very similar. If it's the same or very similar, I'd hope we can align to have only one way of expressing this.

Lastly, is this common practices in your domain to call this just param instead of parameter? Otherwise, I think it's preferrable to use the full name.

> * As for perturbation: users are often either interested in either the control run or all of the perturbation runs. For this it would be sufficient to have a `forecast:mode` attribute with values of either `ctrl` or `perturb`. If desired/necessary it could be paired with a `forecast:sample_nr` with the number of the perturbation run. Since we'll not currently use this it can also be omitted for now and added later if there's a usecase for it.
> * The `forecast:level` will also not be implemented/used in our usecase, but could be useful for others.

I'm a bit concerned about adding "theoretical" use cases. We generally take a pratical approach in STAC and only add properties when they are getting implemented and are practically used, i.e. have a backing in implementations. So I'd propose to potentially split this PR up into the parts, i.e. one PR with properties that you actually implement (which we can merge asap) and one PR (or maybe issue) with potential properties up for discussion and potential implementation. It gives others a forum to chime in and give practical feedback. Once we have practical feedback, we can merge and release it, too.

m-mohr reviewed (changes requested):
Apart from the other comments and open questions, this PR needs a JSON Schema update and CHANGELOG entries.

m-mohr commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:reference_datetime  | string | **REQUIRED.** The *reference* datetime: i.e. predictions for times after this point occur in the future. Predictions prior to this time represent 'hindcasts', predicting states that have already occurred. This must be in UTC. It is formatted according to [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6). |
 | forecast:horizon             | string | The time between the reference datetime and the forecast datetime. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT6H` for a 6-hour forecast. |
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions). |
```
It might be worth it to collection potential parameters in a table and make it a recommended best practice to use certain values for them. The more we can unify this, the more helpful this will be in the long-term. If there are conventions (e.g. in CF or so), we should reuse them.

m-mohr replied:
Result from the discussion with @boecklic:

- Rename to parmeter
- Recommand usage of https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html
- Keep as string

m-mohr replied:
Suggestion: 

```suggestion
| forecast:parameter           | string | Name of the model parameter that corresponds to the data. The parameters should correspond to the [CF Standard Names](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html), e.g. `air_temperature` for the air temperature. |
```

Also, we can add below:

> `forecast:parameter` is primarily intended for search and as such can only be set to a single value, either on the Asset level or on the Item property level. If this field is insufficient for your usecase, please use the more advanced [CF extension](https://github.com/stac-extensions/cf?) instead.

m-mohr commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:horizon             | string | The time between the reference datetime and the forecast datetime. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT6H` for a 6-hour forecast. |
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions). |
+| forecast:mode                | string | Denotes whether the data corresponds to the control run or perturbed runs. Allowed values are `ctrl` or `perturb`. |
```
Is that common practice to use these short terms? Otherwise it might be a bit more intuitive to use `control` and `preturbed`?!

m-mohr replied:
Check for naming conventions, otherwise looks good.

m-mohr replied:
I just saw that ECMWF also names control as "unperturbed".

> The Ensemble Prediction System (EPS) runs 50 different perturbed forecasts with perturbed initial conditions and a control (unperturbed) forecast with a forecast range of 15 days.

See https://www.ecmwf.int/en/forecasts/dataset/operational-archive

So I'm just wondering whether we should actually use a boolean:

```suggestion
| forecast:preturbed           | boolean | Denotes whether the data corresponds to the control run (`false`) or perturbed runs (`true`). The property needs to be specified in both cases as no default value is specified and as such the meaning is "unknown" in case it's missing. |
```

m-mohr commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions). |
+| forecast:mode                | string | Denotes whether the data corresponds to the control run or perturbed runs. Allowed values are `ctrl` or `perturb`. |
+| forecast:member              | integer | Specifies the member (sample number) of perturbed runs, e.g. `1`. |
+| forecast:level               | integer | Index of the vertical level of the height coordinate system used in the forecast model, e.g. `4`. |
```
Do we know which height coordinate system is used? Otherwise, this value seems pretty ambiguous.

m-mohr replied:
Result from the discussion with @boecklic:

Move to separate PR.

Either add a property to set the CRS or add a link relation type that points to documentation that defines these levels.

m-mohr replied:
Moved to #13

m-mohr commented on README.md:
```diff
@@ -30,6 +30,10 @@ The fields in the table below can be used in these parts of STAC documents:
 | forecast:duration            | string | If the forecast is not only for a specific instance in time but instead is for a certain period, you can specify the length here. Formatted as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT3H` for a 3-hour accumulation. If not given, assumes that the forecast is for an instance in time as if this was set to `PT0S` (0 secon ...
+| forecast:param               | string | Name of the model parameter that corresponds to the data, e.g. `T` (temperature), `P` (pressure), `U`/`V`/`W` (windspeed in three directions). |
+| forecast:mode                | string | Denotes whether the data corresponds to the control run or perturbed runs. Allowed values are `ctrl` or `perturb`. |
+| forecast:member              | integer | Specifies the member (sample number) of perturbed runs, e.g. `1`. |
```
Result from the discussion with @boecklic:

Move to separate PR.
Only applies if preturbed is true, implies 0 for pretured = false. That means minimum is 1.

m-mohr replied:
Moved to #13

boecklic committed:
renamed fields and updated their description

boecklic committed:
removed fields that are currently not necessary

boecklic committed:
updated json schema and example

m-mohr commented on json-schema/schema.json:
```diff
@@ -152,7 +152,7 @@
+      "$comment": "Please list all fields here so that we can force the existence of one of them in other parts of the schemas.",
       "anyOf": [
         {"required": ["forecast:reference_datetime"]},
         {"required": ["forecast:horizon"]},
```
```suggestion
        {"required": ["forecast:horizon"]},
        {"required": ["forecast:parameter"]},
        {"required": ["forecast:preturbed"]},
```

rqthomas commented:
It is great to see this extension moving along! Sorry to be late jumping into the discussion, but I wanted to clime in about a few of the suggested updates that relate to how uncertainty is represented in the forecast extension:

The `param` variable currently has the definition of " Name of the model parameter that corresponds to the data."  When I think about a `parameter,` I think about an element of the model (like a scaling coefficient).  I think the term "variable" may be more appropriate.  

Part of the reason to shift from "parameter" to "variable" is to reflect a recent forecasting convention that the Ecological Forecasting Initiative published (https://doi.org/10.1002/ecs2.4686), which considered how to express forecast uncertainty from both an ensemble forecast and distributional forecast (like a mean and sd for a normal distribution).  As a result, we have a column named `family` that is the type of distribution (or ensemble) and parameter (which are the parameters for the distribution of the ensemble member id).  Here is the convention that we follow:

```
| *family*             | Describes the distribution of the prediction (e.g., `normal` for a normal distribution). For ensemble predictions, use `ensemble`. If the predictions yield a single realization (i.e., there is no associated uncertainty), it is preferred to specify `ensemble` with the ensemble size set to 1. If this dimension remains constant, it is acceptable to define *family* as a variable attribute, provided that the file format supports this (e.g., netCDF). Refer to [appendix S1](https://esajournals.onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1002%2Fecs2.4686&file=ecs24686-sup-0001-AppendixS1.pdf) of Dietze et al. 2023 for supported _family_ and _parameter_ names.                                                                                   |
| *parameter*          | For ensemble predictions, specify integers from `1` to `Ne` (where `Ne` represents the total size of the ensemble). For named distributions, indicate the parameter or statistic being specified (e.g., a normal distribution would have `mean` and `sigma` as the values in the *parameter* column).                                                                                                                                                                                                                                         |
| *variable*           | Standardized variable name of what is being predicted. We recommend using [CSDMS Standard Names](https://csdms.colorado.edu/wiki/CSDMS_Standard_Names). For example, the maximum stream temperature would be named `channel_water_surface_water__max_of_temperature.` However, existing widely-used variable names (e.g., `max_temp_c`) are also acceptable.                                                                                                   |
| *prediction*         | Predicted value for the parameter in the *parameter* column. For an ensemble forecast, the value is the prediction for the ensemble member. For a distribution forecast, the value corresponds to the parameter associated with the distribution in the *parameter* column.                                                                                                                                                                                                                                                                                                                                                                                                        |
```

What would be the interest in shifting the conventions here to support both the ensemble-based and distributional forecast output?

m-mohr commented:
I'd be fine with changing the name to variable (or family, although the latter seems less intuitive for me). Any thoughts @boecklic?

@rqthomas I'm not sure which convention to use for the variable though. Is CSDMS Standard Names or CF Standard names more commonly used or what are the differences? Why do two contentions exist in the first place? ;-)

> What would be the interest in shifting the conventions here to support both the ensemble-based and distributional forecast output?

What would that mean effectively for this extension? I'm not a domain expert so would need more input here.

PS: We moved two of the fields to a separate PR for further discussions: https://github.com/stac-extensions/forecast/pull/13

m-mohr reviewed (changes requested):
This looks much better already, I think we are close to merging this. 

We still need to:
- [x] Fix the CI issues
- [x] Apply the suggested change to the schema
- [ ] Discuss the latest comments

rqthomas commented:
@m-mohr `family` describes the type of forecast (e.g., a normal distribution or ensemble forecast) and does not replace `variable`.   The `variable` is the thing that is being forecasted (like air temperature).

Shifting the convention to include uncertainty would mean having the `family` + `parameter` fields as options so that a forecast can optionally be described by its distribution (e.g., a normal distribution) - otherwise the convention would only be allowing "ensemble" forecasts.  The proposal can co-exist with the `forecast:member` field in #12.

It seems that the weather forecast use-case (with the idea of a perturbed ensemble) is well worked through in the proposal.  My suggestions are designed to help the extension include additional use-cases that we encounter in ecology and other environmental sciences, where forecasts may be produced by statistical models that provide the statistical moments (e.g., mean and sd) of a forecast.

boecklic committed:
renamed attribute to resolve conflicting use of parameter

the attribute parameter is used in other forecast scenarios
differently and hence it was suggested and accepted to rename it
to variable.

boecklic committed:
updated changelog

boecklic commented:
Thanks for your input @rqthomas! I'm perfectly fine renaming `parameter` to `variable`, I've already changed it. As for the remaining suggested additional attributes: I'm not a domain expert either, so I cannot really judge, but from your description it makes sense to me.

@m-mohr I've fixed the schema and I hope also the CI (running it needs your approval) and added a changelog entry.

Open point is which convention to recommend in the description of the `variable` attribute.

m-mohr commented on CHANGELOG.md:
```diff
@@ -16,9 +16,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+### Added
+
+- added additional attributes on item/asset level to allow more fine grained filtering of objects
```
Ideally list the property names as well

boecklic replied:
done

m-mohr reviewed (approved):
LGTM

m-mohr commented:
If there's no further feedback from @cboettig @rqthomas @aaronspring or anyone else, I'd merge this PR on Thu or Fri. :-)
Thanks everyone!

boecklic commented:
:tada: thanks for the effort, your inputs and help!
